PhysicsTools/PyTorchAlpaka/interface/QueueGuard.h

```diff
@@ -0,0 +1,119 @@
+#ifndef PhysicsTools_PyTorchAlpaka_interface_QueueGuard_h
+#define PhysicsTools_PyTorchAlpaka_interface_QueueGuard_h
+
+#include <type_traits>
+
+#include <alpaka/alpaka.hpp>
+#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
+#include <cuda_runtime.h>
+#include <c10/cuda/CUDAStream.h>
+#elif ALPAKA_ACC_GPU_HIP_ENABLED
+// #include <c10/hip/HIPStream.h>
+#endif
+
+#include "PhysicsTools/PyTorchAlpaka/interface/alpaka/Config.h"
+
+namespace cms::torch::alpakatools {
+
+  // Default no-op implementation for platforms where no special handling is needed.
+  // CPU backends (do not need extra handling - multithreading is disabled by `PyTorchService`):
+  // - ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED
+  // - ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLED
+  // GPU backends:
+  // - ALPAKA_ACC_GPU_HIP_ENABLED (AMD ROCm/HIP backend not yet supported, see below)
+  template <typename TQueue>
+    requires ::alpaka::isQueue<TQueue>
+  struct QueueGuardTraits {
+    static void set(const TQueue &) noexcept { /* no-op default, threading disabled by `PyTorchService` */ }
+    static void reset() noexcept { /* no-op default, once threading is disabled cannot be reset */ }
+  };
+
+  template <typename TQueue>
+    requires ::alpaka::isQueue<TQueue>
+  class QueueGuard {
+  public:
+    explicit QueueGuard(const TQueue &queue) { QueueGuardTraits<TQueue>::set(queue); }
+    ~QueueGuard() { QueueGuardTraits<TQueue>::reset(); }
+  };
+
+#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
+
+  template <>
+  struct QueueGuardTraits<alpaka_cuda_async::Queue> {
+    inline static thread_local c10::cuda::CUDAStream cached_stream_ = c10::cuda::getCurrentCUDAStream();
+    // setCurrentCUDAStream() is assumed to not throw exceptions on the later-than-first calls.
+    // see: https://github.com/pytorch/pytorch/blob/v2.6.0/c10/cuda/CUDAStream.cpp#L373
+    // Internal torch implementation of CUDA stream handling is based on a `thread_local`
+    // see: https://github.com/pytorch/pytorch/blob/v2.6.0/c10/cuda/CUDAStream.cpp#L169
+    // follows the semantics of "current device" of CUDA itself (but not of Alpaka)
+    //
+    // TODO: `noexcept` is used to avoid exceptions in the destructor, which for 100% clarity
+    // restore the previous state. Not required for correctness and could be neglected due to override behavior.
+    static void set(const alpaka_cuda_async::Queue &queue) noexcept {
+      auto dev = ALPAKA_ACCELERATOR_NAMESPACE::torch::getDevice(queue);
+      auto stream = c10::cuda::getStreamFromExternal(queue.getNativeHandle(), dev.index());
+      cached_stream_ = c10::cuda::getCurrentCUDAStream();
+      c10::cuda::setCurrentCUDAStream(stream);
+    }
+    static void reset() noexcept { c10::cuda::setCurrentCUDAStream(cached_stream_); }
```

Comment:
Is `setCurrentCUDAStream` thread-safe ?
That is, does c10 / Torch internally maintain a different CUDA stream per thread ?

Reply:
Yes, PyTorch tracks the "current CUDA stream" per-thread, using a `thread_local` array of streams indexed by device. See [c10::cuda::CUDAStream.cpp](https://github.com/pytorch/pytorch/blob/v2.6.0/c10/cuda/CUDAStream.cpp#L169). It follows the semantics of "current device" of CUDA itself (but not of Alpaka), and as mentioned in the comment, is assumed not to throw exceptions on the later-than-first calls since there is no heavy logic associated with stream change

Reply:
OK, thanks.